File: app/src/main/java/com/elroid/wirelens/util/FuzzyTextUtils.java
package com.elroid.wirelens.util;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import androidx.collection.ArrayMap;
import androidx.core.util.Pair;
import me.xdrop.fuzzywuzzy.FuzzySearch;
import me.xdrop.fuzzywuzzy.model.ExtractedResult;
import timber.log.Timber;

/**
 * Class: com.elroid.wirelens.util.FuzzyTextUtils
 * Project: WireLens
 * Created Date: 24/01/2018 09:28
 *
 * @author <a href="mailto:e@elroid.com">Elliot Long</a>
 *         Copyright (c) 2018 Elroid Ltd. All rights reserved.
 */
@SuppressWarnings({"WeakerAccess", "SameParameterValue", "unused"})
//this is a util class, methods should be public and may not yet be used
public class FuzzyTextUtils
{
	public static Pair<String, Integer> getClosestMatch(String toMatch, String[] candidates){
		ExtractedResult result = FuzzySearch.extractOne(toMatch, Arrays.asList(candidates));
		return new Pair<>(result.getString(), result.getScore());
	}

	public interface NameExtractor<T>
	{
		String getName(T obj);
	}

	public static <T> List<T> sortListBySimilarity(String toMatch, List<T> list){
		return sortListBySimilarity(toMatch, list, null);
	}

	public static <T> List<T> sortListBySimilarity(String toMatch,
												   List<T> list,
												   NameExtractor<T> extractor){
		return sortListBySimilarity(toMatch, list, extractor, 0, list.size());
	}

	public static <T> List<T> sortListBySimilarity(String toMatch,
												   List<T> list,
												   NameExtractor<T> extractor,
												   int minScore,
												   int maxLen){
		if(extractor == null)
			extractor = obj -> (obj == null) ? "" : obj.toString();

		List<String> names = new ArrayList<>(list.size());
		ArrayMap<String, T> listHash = new ArrayMap<>(list.size());
		for(int i = 0; i < list.size(); i++){
			T t = list.get(i);
			String name = extractor.getName(t).toUpperCase();
			names.add(name);
			listHash.put(name, t);
		}
		List<ExtractedResult> sorted = FuzzySearch.extractSorted(toMatch.toUpperCase(), names);
		List<T> result = new ArrayList<>(list.size());
		if(maxLen == 0) maxLen = list.size();
		int MAX_LEN = Math.min(sorted.size(), maxLen);
		for(int i = 0; i < MAX_LEN; i++){
			ExtractedResult extractedResult = sorted.get(i);
			if(extractedResult.getScore() >= minScore){
				/*Timber.d("result[%s]: string:%s score:%s", i,
					extractedResult.getString(), extractedResult.getScore());*/
				result.add(listHash.remove(extractedResult.getString()));
			}
			else
				break;
		}
		return result;
	}

	public static boolean matches(String str1, String str2, int minSimilarity){
		int sim = FuzzySearch.ratio(str1.toUpperCase(), str2.toUpperCase());
		Timber.v("matches(%s, %s): %s", str1, str2, sim);
		return sim >= minSimilarity;
	}

	private static class CharAlt
	{
		private char[] alts;

		public CharAlt(char... alts){
			this.alts = alts;
		}
	}

	private static List<Character> allAlts;
	private static ArrayMap<Character, CharAlt> altMap;

	static{
		List<CharAlt> charAlts = new ArrayList<>(20);
		charAlts.add(new CharAlt('O', 'D', '0'));
		charAlts.add(new CharAlt('1', 'I', 'l'));
		charAlts.add(new CharAlt('4', 'A'));
		charAlts.add(new CharAlt('5', 'S'));
		charAlts.add(new CharAlt('6', 'G'));
		charAlts.add(new CharAlt('8', 'B'));
		charAlts.add(new CharAlt('M', 'H'));
		allAlts = new ArrayList<>();
		altMap = new ArrayMap<>();
		for(int i = 0; i < charAlts.size(); i++){
			CharAlt charAlt = charAlts.get(i);
			for(char c : charAlt.alts){
				allAlts.add(c);
				altMap.put(c, charAlt);
			}
		}
	}

	public static List<String> findSimilarLookingWords(String word){
		List<String> result = new ArrayList<>();
		result.add(word);
		for(int i = 0; i < word.length(); i++){
			char c = word.charAt(i);
			if(allAlts.contains(c)){
				List<String> similarsForIndex = new ArrayList<>();
				for(String simWord : result){
					List<String> similars = findSimilarLookingWords(simWord, i);
					//Timber.d("found similar(%s) to %s: %s", i, simWord, similars);
					similarsForIndex.addAll(similars);
				}
				result.addAll(similarsForIndex);
			}
		}
		return result;
	}

	/*private static List<String> removeDuplicates(List<String> list){
		List<String> result = new ArrayList<>(list.size());
		Timber.d("pre-dup :%s", list);
		for(String s : list){
			if(!result.contains(s))
				result.add(s);
		}
		Timber.d("post-dup:%s", result);
		return result;
	}*/

	private static List<String> findSimilarLookingWords(String word, int index){
		char c = word.charAt(index);
		CharAlt alt = altMap.get(c);
		List<String> result = new ArrayList<>(alt.alts.length);
		for(int i = 0; i < alt.alts.length; i++){
			char a = alt.alts[i];
			String sim = setCharAt(word, index, a);
			if(!word.equals(sim))
				result.add(sim);
		}
		return result;
	}

	private static String setCharAt(String word, int index, char c){
		return word.substring(0, index) + c + word.substring(index + 1);
	}

}

File: app/src/main/java/com/elroid/wirelens/util/FuzzyTextUtils.kt
package com.elroid.wirelens.util

import me.xdrop.fuzzywuzzy.FuzzySearch
import timber.log.Timber
import java.util.Locale
import kotlin.math.min

/**
 * Class: com.elroid.wirelens.util.FuzzyTextUtils
 * Project: WireLens
 * Created Date: 24/01/2018 09:28
 *
 * @author [Elliot Long](mailto:e@elroid.com)
 * Copyright (c) 2018 Elroid Ltd. All rights reserved.
 */
object FuzzyTextUtils {

	private val allAlts = mutableListOf<Char>()
	private val altMap = mutableMapOf<Char, CharAlt>()

	private class CharAlt(vararg val alts: Char)

	init {
		val charAlts = mutableListOf(
			CharAlt('O', 'D', '0'),
			CharAlt('1', 'I', 'l'),
			CharAlt('4', 'A'),
			CharAlt('5', 'S'),
			CharAlt('6', 'G'),
			CharAlt('8', 'B'),
			CharAlt('M', 'H')
		)
		charAlts.forEach { charAlt ->
			charAlt.alts.forEach {
				allAlts.add(it)
				altMap[it] = charAlt
			}
		}
	}

	/*fun getClosestMatch(toMatch: String?,
						candidates: Array<String?>): Pair<String, Int> {
		val result =
			FuzzySearch.extractOne(toMatch, Arrays.asList<String>(*candidates))
		return Pair(result.string, result.score)
	}*/

	fun <T> sortListBySimilarity(toMatch: String,
								 list: List<T>,
								 minScore: Int = 0,
								 maxLen: Int = list.size,
								 extractor: (T) -> String = { obj ->
									 obj?.toString() ?: ""
								 }): List<T> {
		val names = mutableListOf<String>()
		val listHash = mutableMapOf<String, T>()
		for(i in list.indices) {
			val t = list[i]
			val name = extractor(t).toUpper()
			names.add(name)
			listHash[name] = t
		}
		val sorted = FuzzySearch.extractSorted(toMatch.toUpper(), names)
		val result = mutableListOf<T>()
		val maxLength = min(sorted.size, maxLen)
		for(i in 0 until maxLength) {
			val extractedResult = sorted[i]
			Timber.d("result[%s]: string:%s score:%s", i,
				extractedResult.string, extractedResult.score)
			if(extractedResult.score >= minScore) {
				listHash.remove(extractedResult.string)?.let {
					result.add(it)
				}
			} else break
		}
		return result
	}

	private fun String.toUpper(): String = toUpperCase(Locale.getDefault())

	@JvmStatic
	fun matches(str1: String, str2: String, minSimilarity: Int): Boolean {
		val sim = FuzzySearch.ratio(str1.toUpper(), str2.toUpper())
		Timber.v("matches(%s, %s): %s", str1, str2, sim)
		return sim >= minSimilarity
	}

	fun findSimilarLookingWords(word: String): List<String> {
		val result = mutableListOf<String>()
		result.add(word)
		for(i in word.indices) {
			val c = word[i]
			if(allAlts.contains(c)) {
				val similarWordsForIndex = mutableListOf<String>()
				for(simWord in result) {
					val similarWords = findSimilarLookingWords(simWord, i)
					//Timber.d("found similar(%s) to %s: %s", i, simWord, similars);
					similarWordsForIndex.addAll(similarWords)
				}
				result.addAll(similarWordsForIndex)
			}
		}
		return result
	}

	private fun findSimilarLookingWords(word: String,
										index: Int): List<String> {
		val c = word[index]
		val alt = altMap[c]
		val result = mutableListOf<String>()
		alt?.alts?.forEach {
			val sim = setCharAt(word, index, it)
			if(word != sim && !result.contains(sim))
				result.add(sim)
		}
		return result
	}

	private fun setCharAt(word: String, index: Int, c: Char): String {
		return word.substring(0, index) + c + word.substring(index + 1)
	}

}